src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -62,29 +64,31 @@ def __init__(
     def resolve(self, root_reqs, check_supported_wheels):
         # type: (List[InstallRequirement], bool) -> RequirementSet
 
-        # The factory should not have retained state from any previous usage.
-        # In theory this could only happen if self was reused to do a second
-        # resolve, which isn't something we do at the moment. We assert here
-        # in order to catch the issue if that ever changes.
-        # The persistent state that we care about is `root_reqs`.
-        assert len(self.factory.root_reqs) == 0, "Factory is being re-used"
-
-        # FIXME: Implement constraints.
-        if any(r.constraint for r in root_reqs):
-            raise InstallationError("Constraints are not yet supported.")
+        constraints = defaultdict(list)  # type: Dict[str,List[SpecifierSet]]
+        requirements = []
+        roots = set()  # type: Set[str]
+        for req in root_reqs:
+            if req.constraint:
+                assert req.name
+                assert req.specifier
+                name = canonicalize_name(req.name)
+                constraints[name].append(req.specifier)
```

Comment:
I wonder if it is possible to merge the specifiers here. It would make the contraint checking code more readable since the loops can be eliminated.